README
Autor: João S. O. Bueno

Seguindo discussões nas listas python brasile  python-gamedev-brasil da semana de 
25 de agosto de 2008, resolvi fazer esta implementação com os seguintes intuitos:
    - usar na minah apresnetação no evento da impacta dia 30/08
    http://www.impacta.com.br/eletter/palestra_python/eletter.asp - em 29/08/2008
    
O código está bem bobinho, só paramostrar um exmeplo mesmo sobre o qual pretendo falar amanhã.
Se eu tiver disposição hoje ainda, vou aproveitar que fiz o jogo de forma bastante abstraida e fazer um front-end em pygame também, alem do atual usando pyglet.


OS demais usuários do Pygame-devel, fiquem a  vontade para evoluir o projeto - 
inclusive como a pversãoq eu estou faznedo commit agora foi feita em dois dias
tem umas coisas meio "add hoc" (vulgo: gambiarras): tipo coiassa similares, feitas
em lugares diferentes de jeitos diferentes.

Vou deixar assim pra falar amanhã na palestra, mas depois é legal

1) unificar o carregamento de mídia (imagem e sons) : estão sendo feitos 
em lugares diferentes e de um jeito um pouco diferente

2) usar um método único de notificação para a Cobra: ela está levantando uma exceção quando morre, e usando um callback quando coem uma maçã. Resolvi que o melhor
mesmo é o callback - e, s eo porjeto for crescer, uma l"lista de callbacks",
que na prática re-implementa "sinais e slots"  dos sistemas de programação GUI.



ChangeLog
2008-08-30  João S. O. Bueno  <gwidion@mpc.com.br>

       * FreeSans.ttf: Added.
       * COPYING: Updated to mention FreeSans.ttf as non GPL 3 file

2008-08-29  João S. O. Bueno  <gwidion@mpc.com.br>

       * pygame_snake.py: Criado - implementacao de pyglet alterada
       para funcioanr com pygame.
       * TODO: Criado

2008-08-29  João S. O. Bueno  <gwidion@mpc.com.br>

        * snake.py: corrige para que a instância do mapa
        passe a existir quando executa o teste em modo texto.

2008-08-29  João S. O. Bueno  <gwidion@mpc.com.br>

        * snakes.py
        * pyglet_snakes.py
        * AUTHORS
        * KDE_Beep_Bottles.wav
        * COPYING
        * KDE_Critical_Error.wav
        * esfera_1.png, parede.png, quadrado_preto.png, pop.wav:
        Atribuição de licença GPL v3.0 aos projetos e commit inicial
        no google-code da python-gamedev-brasil

2008-08-28  João S. O. Bueno  <gwidion@mpc.com.br>

        * snake.py: correção de bug no algorítmo de extensão da cobra;
        várias atualizações para funcionar como base para a parte gráfica

        * pyglet_snake.py: impleemtnação inicial da parte gráfica com pyglet,
        com suporte a imagens, controle por teclado, som, contagem de pontos,
        exibição de placar simples quando a cobra morre e reinício do jogo


2008-08-27  João S. O. Bueno  <gwidion@mpc.com.br>

        * snake.py  - implementação inicial do mapa e da cobra, em modo texto
          a partir de e-mail enviado por Roberto para a lista
          python-gamedev-brasil@googlegroups.com



snake.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#SNAKE - jogo clássico em python, implementação pedagógica
#Copyright (C) 2008  <João S. O. Bueno>, <Roberto Fagá>

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <http://www.gnu.org/licenses/>.


#Authors: Roberto Fagá (Portuguese),
# João S. O. Bueno (translation to Python)

from array import array
import sys

#sequencia ANSIpara limpar o terminal e posicionar o ursor no inicio:
#FIXME: isso não funciona pro padrão no windows -
#para os testes, deve-se achar um jeito "barato" de limpar
#o terminal de texto.
HOME_ANSI = "\x1b[2J\x1b[0;0H"
#J. Exceções para tornar mais simples saber quando a cobra morreu
class CobraMorreu(Exception):
    pass

class CobraBateuEmSiMesma(CobraMorreu):
    pass

class CobraBateuNaParede(CobraMorreu):
    pass


#R. Usar uma matriz grande

LARGURA, ALTURA = GRANDE = (30, 30)

#R. Cada valor da matriz dessa forma é um valor numérico, por exemplo,
#R. e o número remete ao código da imagem e o objeto correspondente.
#R. Ex: de 1 a 3 é a cobra (cabeça, corpo ou calda),

VAZIO = 0
COBRA_CABECA = 1
COBRA_CORPO = 2
COBRA_CAUDA = 3

#R. 5 é a maçã,
MACA = 5

#R. 8 é  parede, algo assim.
# Faça sempre com constantes nesse caso.

PAREDE = 8



#R. (pode ser lista dentro de lista mesmo) em
#R. que cada ponto pode ser parede, item a pegar ou parte da cobra.


class Mapa(object):
    from array import array
    def __init__(self, largura=LARGURA, altura=ALTURA):
        self.largura, self.altura = largura, altura
        self.data = array("B", [VAZIO  for i in xrange(largura * altura)])

    def __getitem__(self, indice):
        return self.data[indice[0] + indice[1] * self.largura]

    def __setitem__(self, indice, content):
        if indice is None:
            return
        try:
            self.data[indice[0] + indice[1] * self.largura] = content
        except IndexError:
            pass

    def linha_h(self, linha, x1, x2):
        for x  in range(x1, x2):
            self[x, linha] = PAREDE
    def linha_v(self, coluna, y1, y2):
        for y  in range(y1, y2):
            self[coluna, y] = PAREDE
    def retangulo(self, x1, y1, x2, y2):
        self.linha_h(y1, x1, x2 - 1)
        self.linha_h(y2 - 1, x1, x2 - 1)
        self.linha_v(x1, y1, y2- 1)
        self.linha_v(x2 - 1, y1, y2 - 1)
        self[x2 - 1, y2 - 1] = PAREDE

    def desenha(self):
        """Imprime mapa no terminal para testes -
           deve ser overrided numa sub-classe para
           implementacção gráfica
        """

        dic_figuras = {PAREDE:"+", COBRA_CORPO:"*", COBRA_CABECA:"@", COBRA_CAUDA:".", VAZIO:" "}
        sys.stdout.write(HOME_ANSI)
        for y in range(self.altura):
            for x in range(self.altura):
                sys.stdout.write(dic_figuras[self[x,y]] + " ")
            sys.stdout.write("\n")

itens_mortais = (PAREDE, COBRA_CABECA, COBRA_CORPO, COBRA_CAUDA)

#R. Para o
# movimento da cobra aqui é bem simples, vc simplesmente armazena a
# direção em uma variável +1 a -1 na horizontal e na vertical. Dessa
# forma, se o usuário pressionar "direita" por exemplo, essas
# variáveis assumem valor horizontal +1 e vertical 0.

class Cobra(list):
    def __init__(self, posicao, direcao, velocidade, comprimento, mapa = None):
        """J. posicao => tupla de coordenadas da cobra no mapa
            direcao -> tupla com a direcao x e y de movimento da cobra -
                       cada componente podendo ser 1, 0 ou -1
            velocidade -> numero de interacoes do laço principal entre
                         as atualizações da  posição da cobra
        """
        list.__init__(self)
        self.posicao = posicao
        self.direcao = direcao
        self.velocidade = velocidade
        x, y = tuple(posicao)
        for pedaco in range(comprimento):
            self.append((x,y))
            #j. o corpo se extende na direção contrária ao movimento
            x -=  direcao[0]
            y -=  direcao[1]
        self.cabeca = 0
        self.aumentando = 0
        self.pos_rastro = None
        if mapa:
            self.mapeia(mapa)
        self.contador = 0
        self.chama_maca = None

    def atualiza(self):
        self.contador += 1
        if not self.contador % self.velocidade:
            self.mover()

    #J. definimos os metodos de acesso aos itens, de modo que
    # o acesso a coordenadas do corpo da cobra seja normalizado
    # em relação a cabeça. (i.e. cobra[0] sempre vai retornar
    # as coordenadas da cabeça)
    def __getitem__(self, posicao):
        return list.__getitem__(self, (self.cabeca + posicao) % len(self))
        #J. O divertido é que só isso funciona inclusive para itens negativos.
        # cobra[-1] vai retornar as coordenadas da cauda

    def __setitem__(self, posicao, valor):
        list.__setitem__(self, (self.cabeca + posicao) % len(self), valor)

    def __iter__(self):
        def cobraIter(cobra):
            indice = 0
            while True:
                if indice == len(cobra):
                    raise StopIteration
                indice += 1
                yield cobra[indice - 1]
        return cobraIter(self)

    def insert(self, posicao, valor):
        list.insert(self, (self.cabeca + posicao) % len(self), valor)

    #R. A cada refresh
    # de tela a posição da cobra é então incrementada, fazendo com que
    # cada casa dela desloque para a casa que estava a parte anterior da
    # cobra e a cabeça na direção nova. Vc pode armazenar em um vetor a
    # cobra, sendo que cada casa do vetor é a posição atual da matriz,
    # assim pode ficar mais fácil de dar o refresh.
    def mover(self):
        x, y = self[0]
        x += self.direcao[0]
        y += self.direcao[1]

        #J. Se a cobra está em fase de crescimento, aumentar o vetor
        if self.aumentando > 0:
            self.esta_aumentando = True
            self.insert(0, None)
            self.aumentando -= 1
        else:
            self.esta_aumentando = False
            self.cabeca -= 1
            if self.cabeca < 0:
                self.cabeca = len(self) - 1

            self.pos_rastro = self[0]
        self[0] = None
        self.testa_morte_so((x,y))
        self.testa_maca((x,y))
        self[0] = (x,y)
        if hasattr(self, "mapa"):
            self.atualiza_mapa()


    #R. O teste de colisão é
    # muito simples, é só testar se no refresh, a cabeça foi para uma
    # casa já ocupada ou não.
    def testa_morte_so(self, pos):
        if pos in self:
            raise CobraBateuEmSiMesma("bum")

    def testa_morte_mapa(self, pos):
        if self.mapa[pos] == PAREDE:
            raise CobraBateuNaParede("Poft!")
        elif self.mapa[pos] in itens_mortais:
                raise CobraMorreu("argh")

    def testa_maca(self, pos):
        if self.mapa[pos] == MACA:
            self.chama_maca(pos)

    def mapeia(self, mapa):
        """J. Aplica a cobra num mapa limpo, e associa este mapa
           à instancia da cobra
        """
        self.mapa = mapa
        dic_parte = {0: COBRA_CABECA, len(self) -1: COBRA_CAUDA}
        for i, pedaco in enumerate(self):
            self.testa_morte_mapa(pedaco)
            self.mapa[pedaco] = dic_parte.get(i, COBRA_CORPO)

    def atualiza_mapa(self):
        self.testa_morte_mapa(self[0])
        if not self.esta_aumentando: #self.pos_rastro is not None :
            self.mapa[self.pos_rastro] = VAZIO
            self.mapa[self[-1]] = COBRA_CAUDA
        self.mapa[self[0]] = COBRA_CABECA
        self.mapa[self[1]] = COBRA_CORPO
        #if self.esta_aumentando:
        #    self.mapa[self[2]] = COBRA_CORPO
            #self.mapa[self[-2]] = COBRA_CORPO

def mapa_0(classe_mapas=Mapa,  *args, **kwargs):
    mapa = classe_mapas(*args, **kwargs)
    mapa.retangulo(0, 0, mapa.largura, mapa.altura)
    return mapa

if __name__ == "__main__":
    from time import sleep

    mapa = mapa_0()
    #mapa.linha_h(5, 10, 20)
    cobra = Cobra((4,3), (1,0), 6, 4, mapa)

    mapa.desenha()
    t = 0.1
    for i in range(60):
        cobra.mover()
        mapa.desenha()
        sleep(t)
        if i == 10:
            cobra.direcao=(0,1)
        elif i == 20:
            cobra.aumentando = 3
        elif i == 30:
            cobra.direcao=(-1,0)
            cobra.aumentando += 4
        elif i == 37:
            cobra.direcao=(0, -1)
        elif i == 48:
            cobra.direcao = (1,0)


# Você pode até fazer efeito de animação, colocando o incremento de
# -0.25 a +0.25 por exemplo, e só aceitando o controle e testando
# colisão de 4 em 4 quadros. A maçã faz crescer a cobra, é só
# aumentar o vetor da cobra, apagar a maçã e randomizar uma nova em
# algum lugar do mapa desocupado (pode ser por tentativa e erro
# mesmo).
# Depois esses mapas, sem a cobra e a maçã, podem ser armazenados em
# arquivos com cPickle mesmo, que armazena variáveis Python. Aí vc só
# lê a matriz do mapa e coloca a cobra e a maçã.


pyglet_snake.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#SNAKE - jogo clássico em python, implementação pedagógica
#Copyright (C) 2008  <João S. O. Bueno>

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <http://www.gnu.org/licenses/>.


#Authors:  João S. O. Bueno (translation to Python)

import pyglet
from copy import copy
from random import randrange

import snake
from snake import PAREDE, VAZIO, COBRA_CORPO, COBRA_CABECA, COBRA_CAUDA, MACA

class GletMapa(snake.Mapa):
    def __init__(self, janela, largura, altura):
        self.janela = janela
        snake.Mapa.__init__(self, largura, altura)
        self.larg_celula = janela.width // self.largura
        self.alt_celula = janela.height // self.altura
        self.carrega_imagens()
        self.mapa_anterior = snake.Mapa(largura,altura)
        self.counter = 0
    def carrega_imagens(self):
        #l = lambda nome: pyglet.sprite.Sprite(pyglet.image.load(nome))
        l = pyglet.image.load
        self.imagens = {PAREDE: l("parede.png"),
                        VAZIO: l("quadrado_preto.png"),
                        COBRA_CORPO: l("esfera_1.png"),
                        COBRA_CABECA: l("esfera_2.png"),
                        COBRA_CAUDA: l("esfera_1.png"),
                        MACA: l("esfera_2.png")
                       }
        #for imagem in self.imagens.values():
            #imagem.scale = self.larg_celula / float(imagem.width)
            #print imagem.scale

    #def coord_mapa_pra_janela(self, (x, y)):
    #    return (x / float(mapa.largura) * janela.width,
    #            y / float(mapa.altura) * janela.height)
    def mantem_mapa_anterior(self):
        self.mapa_anterior.data = copy(self.data)

    def desenha(self):
        for y in range(self.altura):
            for x in range(self.largura):
                if self.counter != 0 and self[x, y] == self.mapa_anterior[x, y]:
                    continue
                imagem = self.imagens[self[x,y]]
                #print imagem,imagem.width, x, y
                i_x = x * self.larg_celula
                i_y = self.janela.height - ( (y +1 ) * self.alt_celula)
                imagem.blit(i_x, i_y)
        self.mantem_mapa_anterior()
        self.counter += 1

class Game(pyglet.window.Window):
    def __init__(self, largura, altura, *args, **kwargs):
        pyglet.window.Window.__init__(self, largura, altura, *args, **kwargs)
        self.carrega_sons()
        self.inicializa_jogo()

    def inicializa_jogo(self):
        pyglet.clock.unschedule(self.atualiza)
        self.mapa = snake.mapa_0(GletMapa, self, 40, 30)
        self.cobra = snake.Cobra((5,2), (1,0), 2, 4, self.mapa)
        self.cobra.chama_maca = self.comeu_maca
        self.pontos = 0
        self.jogo_acabou = False
        self.nova_maca()
        self.contador = 0
        pyglet.clock.schedule_interval(self.atualiza, 1/20.0)

    def carrega_sons(self):

        self.sons = {"maca_aparece": "pop.wav",
                     "comeu_maca": "KDE_Beep_Bottles.wav",
                     "morreu": "KDE_Critical_Error.wav"}
        for evento, arquivo in self.sons.items():
            self.sons[evento] = pyglet.resource.media(arquivo, streaming=False)


    def atualiza(self, relogio):
        self.contador += 1
        if not self.jogo_acabou:
            try:
                self.cobra.atualiza()
            except snake.CobraMorreu, morreu:
                self.jogo_acabou = True
                self.sons["morreu"].play()
                self.placar = pyglet.text.Label("%d pontos!" % self.pontos,
                                      font_name = "Sans",
                                      font_size = 80,
                                      x = self.width // 2,
                                      y = self.height // 2,
                                      anchor_x = "center",
                                      anchor_y = "center")
        else:
            self.placar.color=(255, 255, 10 * self.contador % 255, 255)
            self.placar.draw()

    def on_draw(self, *args):
        #self.clear()
        self.mapa.desenha()

    def on_key_press(self, simbolo, modificador):
        key = pyglet.window.key
        cobra = self.cobra
        if simbolo == key.RIGHT:
            cobra.direcao = (1,0)
        elif simbolo == key.LEFT:
            cobra.direcao = (-1,0)
        elif simbolo == key.DOWN:
            cobra.direcao = (0,1)
        elif simbolo == key.UP:
            cobra.direcao = (0,-1)
        if simbolo == key.ESCAPE:
            self.close()
        if self.jogo_acabou and simbolo == key.SPACE:
            self.inicializa_jogo()

    def comeu_maca(self, pos):
        self.sons["comeu_maca"].play()
        self.pontos += 1
        pyglet.clock.schedule_once(self.nova_maca, 1)
        #self.nova_maca()
        self.cobra.aumentando += 4

    def nova_maca(self, relogio = 0):
        pos = (0,0)
        while (self.mapa[pos] != VAZIO):
            pos = (randrange(self.mapa.largura),
                   randrange (self.mapa.altura))
        self.sons["maca_aparece"].play()
        self.mapa[pos] = MACA

game = Game(800,600)
pyglet.app.run()

pygame_snake.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#SNAKE - jogo clássico em python, implementação pedagógica
#Copyright (C) 2008  <João S. O. Bueno>

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <http://www.gnu.org/licenses/>.


#Authors:  João S. O. Bueno

import pygame
from copy import copy
from random import randrange

import snake
from snake import PAREDE, VAZIO, COBRA_CORPO, COBRA_CABECA, COBRA_CAUDA, MACA

TELA = None #variável global, mantendo a superficie da Tela do jogo

class GameMapa(snake.Mapa):
    def __init__(self, janela, largura, altura):
        self.janela = janela
        snake.Mapa.__init__(self, largura, altura)
        self.larg_celula = janela.width // self.largura
        self.alt_celula = janela.height // self.altura
        self.carrega_imagens()
        self.mapa_anterior = snake.Mapa(largura,altura)
        self.counter = 0
    def carrega_imagens(self):
        l = pygame.image.load
        self.imagens = {PAREDE: l("parede.png"),
                        VAZIO: l("quadrado_preto.png"),
                        COBRA_CORPO: l("esfera_1.png"),
                        COBRA_CABECA: l("esfera_2.png"),
                        COBRA_CAUDA: l("esfera_1.png"),
                        MACA: l("esfera_2.png")
                       }

    def mantem_mapa_anterior(self):
        self.mapa_anterior.data = copy(self.data)

    def desenha(self):
        for y in range(self.altura):
            for x in range(self.largura):
                if self.counter != 0 and self[x, y] == self.mapa_anterior[x, y]:
                    continue
                imagem = self.imagens[self[x,y]]
                i_x = x * self.larg_celula
                i_y =  y  * self.alt_celula
                TELA.blit(imagem, (i_x, i_y))
        self.mantem_mapa_anterior()
        self.counter += 1

class Game(object):
    def __init__(self, largura, altura, *args, **kwargs):
        global TELA
        TELA = pygame.display.set_mode((largura,altura))
        self.width = self.largura = largura
        self.height = self.altura = altura
        self.carrega_sons()
        self.inicializa_jogo()

    def laco_principal(self):
        while True:
            pygame.time.wait(50)
            eventos = self.agenda.keys()
            while (self.agenda and pygame.time.get_ticks() > min(self.agenda.keys())):
                chave_evento = min(self.agenda.keys())
                self.agenda[chave_evento]()
                del self.agenda[chave_evento]
            pygame.event.pump ()
            eventos = pygame.event.get()
            for evento in eventos:
                if evento.type == pygame.KEYDOWN:
                    self.on_key_press(evento.key)
            self.atualiza()

    def inicializa_jogo(self):
        self.mapa = snake.mapa_0(GameMapa, self, 40, 30)
        self.cobra = snake.Cobra((5,2), (1,0), 2, 4, self.mapa)
        self.cobra.chama_maca = self.comeu_maca
        self.pontos = 0
        self.jogo_acabou = False
        self.nova_maca()
        self.contador = 0
        self.agenda = {}

    def carrega_sons(self):

        self.sons = {"maca_aparece": "pop.wav",
                     "comeu_maca": "KDE_Beep_Bottles.wav",
                     "morreu": "KDE_Critical_Error.wav"}
        class Som(object):
            def __init__(self, arquivo):
                self.som = pygame.mixer.Sound(arquivo)
                self.num_canais = pygame.mixer.get_num_channels()
                self.canais = []
                for i in range(self.num_canais):
                    self.canais.append(pygame.mixer.find_channel(i))
            def play(self):
                ocupado = pygame.mixer.get_busy()
                c = 0
                while True:
                    if not ocupado % 2:
                        break
                    c += 1
                    ocupado >>= 1
                if c < self.num_canais:
                    self.canais[c].play(self.som)

        for evento, arquivo in self.sons.items():
            self.sons[evento] = Som(arquivo)


    def atualiza(self, relogio=0):
        self.contador += 1
        if not self.jogo_acabou:
            try:
                self.cobra.atualiza()
                self.on_draw()
            except snake.CobraMorreu, morreu:
                self.jogo_acabou = True
                self.sons["morreu"].play()
                self.texto_placar = "%d pontos!" % self.pontos
                self.fonte = pygame.font.Font("FreeSans.ttf", 80)


        else:
            cor = (255, 255, 10 * self.contador % 255)
            imagem_placar = self.fonte.render(self.texto_placar, True, cor)
            TELA.blit(imagem_placar,
                      (self.width // 2 - imagem_placar.get_width() //2,
                       self.height // 2 - imagem_placar.get_height() //2))
        pygame.display.flip()

    def on_draw(self, *args):
        #self.clear()
        self.mapa.desenha()

    def on_key_press(self, simbolo, modificador = 0):
        cobra = self.cobra
        if simbolo == pygame.K_RIGHT:
            cobra.direcao = (1,0)
        elif simbolo == pygame.K_LEFT:
            cobra.direcao = (-1,0)
        elif simbolo == pygame.K_DOWN:
            cobra.direcao = (0,1)
        elif simbolo == pygame.K_UP:
            cobra.direcao = (0,-1)
        if simbolo == pygame.K_ESCAPE:
            self.encerrar()
        if self.jogo_acabou and simbolo == pygame.K_SPACE:
            self.inicializa_jogo()


    def comeu_maca(self, pos):
        self.sons["comeu_maca"].play()
        self.pontos += 1
        self.agenda[pygame.time.get_ticks() + 1000] = self.nova_maca
        #self.nova_maca()
        self.cobra.aumentando += 4

    def nova_maca(self, relogio = 0):
        pos = (0,0)
        while (self.mapa[pos] != VAZIO):
            pos = (randrange(self.mapa.largura),
                   randrange (self.mapa.altura))
        self.sons["maca_aparece"].play()
        self.mapa[pos] = MACA

    def encerrar(self):
        pygame.quit()


pygame.init()
game = Game(800,600)
game.laco_principal()